lib/features/modeling/cmd/ToggleShapeCollapseHandler.js

```diff
@@ -16,16 +21,16 @@ ToggleShapeCollapseHandler.prototype.execute = function(context) {
   var shape = context.shape,
       children = shape.children;
 
-  // remember previous visibility of children
-  context.oldChildrenVisibility = getElementsVisibility(children);
+  // recursively remember previous visibility of children
+  context.oldChildrenVisibility = getElementsVisibilityRecursive(children);
 
   // toggle state
   shape.collapsed = !shape.collapsed;
 
-  // hide/show children
-  setHidden(children, shape.collapsed);
+  // recursively hide/show children
+  var result = setHiddenRecursive(children, shape.collapsed);
```

Comment:
Did you consider a CSS only solution to hide the children?

Reply:
No, I have not, but what setHidden method does is to set the hidden property of a shape, then GraphicsFactory set display:none to those having hidden = true. Could you explain further what do you mean by CSS only solution?

Reply:
Given how the elements are structured in the DOM we could create a CSS class, lets say `djs-hidden` on the element group that hides all elements via `display: none` in the hierarchy. 

Not sure if it is a good approach, but maybe one we could consider.

Reply:
![Kapture 2019-10-18 at 10 30 10](https://user-images.githubusercontent.com/15003836/67078813-5a0f9280-f192-11e9-927e-b1639f23154a.gif)

Adding display:none to related djs-children class would be a solution (this does not hide direct editing related DOM elements though but this also can be fixed). However, IMO this creates an inconsistency regarding the Shape class, as the hidden property would still be false/undefined even though the shape itself is hidden. So here, setHiddenRecursive method ensures that the hidden property of all the children/subchildren are correctly set as well.

Reply:
Thanks for investigating. 

> However, IMO this creates an inconsistency regarding the Shape class, as the hidden property would still be false/undefined even though the shape itself is hidden. 

Agreed. We should go for the proper, consistent solution.